music/src/ddsp/ddsp.ts

```diff
@@ -0,0 +1,387 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as Tone from 'tone';
+import * as tf from '@tensorflow/tfjs';
+import { resampleAndMakeMono, hzToMidi } from '../core/audio_utils';
+import { mixAndJoinAudioData } from './audio_utils';
+import {
+  CROSSFADE_DURATION,
+  OUTPUT_SAMPLE_RATE,
+  LOWEST_LD,
+  CONF_SMOOTH_SIZE,
+  CONF_THRESHOLD,
+  LD_CONF_REDUCTION,
+  PRESET_MODELS,
+  MIN_VRAM,
+} from './constants';
+import { AudioFeatures, ModelValues } from './interfaces';
+import { upsample_f0, shiftF0 } from '../spice/pitch_utils';
+import { Tensor3D, Tensor } from '@tensorflow/tfjs';
+import { addReverb } from './add_reverb';
+import { arrayBufferToAudioBuffer } from './buffer_utils';
+import { MODEL_SAMPLE_RATE, MODEL_FRAME_RATE } from '../spice/spice';
+
+async function memCheck() {
+  const bytesPerMB = 1024 * 1024;
+  const screenResToVRAMFactor = 600;
+  const screenSize = window.screen.availWidth * window.screen.availHeight;
+  const DPI = window.devicePixelRatio;
+
+  // checking of vram ensures that the device can hold
+  // all the models and operations
+  const vramSize = Math.round(
+    (screenSize * DPI * screenResToVRAMFactor) / bytesPerMB
+  );
+
+  // 50MB limit
+  if (!isNaN(vramSize) && vramSize < MIN_VRAM) {
+    throw new Error(
+      `Insufficient memory! Your device has ${vramSize} and recommended memory is ${MIN_VRAM}`
+    );
+  }
+
+  // loading a model and then executing it tests if
+  // the device can hold the texture size needed by the model
+  // load the biggest model and test it
+  try {
+    await tf.ready();
+
+    if (tf.getBackend() !== 'webgl') {
+      throw new Error('It looks like your browser does not support webgl.');
+    }
+  } catch (err) {
+    throw new Error(`insufficient memory - ${err}`);
+  }
+
+  return true;
+}
+
+export async function getModel(url: string) {
+  const model = await tf.loadGraphModel(url);
+  return model;
+}
+
+function resizeAudioFeatures(
+  audioFeatures: AudioFeatures,
+  startingFrame: number,
+  endingFrame: number,
+  isLastChunks: boolean,
+  resizedLength: number
+): Promise<AudioFeatures> {
+  return new Promise((resolve) => {
+    const resizedAudioFeatures = {
+      loudness_db: [0],
+      f0_hz: [0],
+    };
+    let counter = 0;
+    for (let i = startingFrame; i < endingFrame; i++) {
+      if (isLastChunks && i >= resizedLength) {
+        resizedAudioFeatures.loudness_db[counter] = LOWEST_LD;
+        resizedAudioFeatures.f0_hz[counter] = -1;
+      } else {
+        resizedAudioFeatures.loudness_db[counter] =
+          audioFeatures.loudness_db[i];
+        resizedAudioFeatures.f0_hz[counter] = audioFeatures.f0_hz[i];
+      }
+      counter += 1;
+    }
+
+    resolve(resizedAudioFeatures);
+  });
+}
+
+async function ddsp_process(
```

Comment:
Same for this one, it only really makes sense in the context of a model so it seems cleaner to just make it a method of the model class.